File: cpp/vendor1.cpp

namespace Vendor1 {
  using namespace soup;

  typedef struct {
    char type;
    int x;
  } Vendor1Device;


  class Vendor1DeviceA : public DeviceA {
    Vendor1Device dev;
  public:
    Vendor1DeviceA(int index) : dev({'A', index}) {}

    virtual void op1(int param) {
      dev.x = param;
    }

    virtual int op2(int param) {
      return ++dev.x;
    }
  };

  class Vendor1DeviceB : public DeviceB {
    Vendor1Device dev;
  public:
    Vendor1DeviceB(int index) : dev({'B', index}) {}

    virtual void op1(int param) {
      dev.x *= param;
    }

    virtual int op2(int x, int y) {
      return dev.x * x + y;
    }
  };


  class Vendor1Vendor : public Vendor {
  public:
    virtual DeviceA* createDeviceA(int index) {
      return new Vendor1DeviceA(index);
    }

    virtual DeviceB* createDeviceB(int index) {
      return new Vendor1DeviceB(index);
    }
  };


  Vendor* GetVendor() {
    return new Vendor1Vendor();
  }

}

File: rust/src/vendor1.rs

use soup;


pub struct Vendor1 {}

struct Vendor1DeviceA {
    x: i32
}

struct Vendor1DeviceB {
    x: i32
}


impl Vendor1DeviceA {
    fn new(index: i32) -> Vendor1DeviceA {
        Vendor1DeviceA { x: index }
    }
}


impl soup::Device for Vendor1DeviceA {

    fn op1(&mut self, param: i32) {
        self.x = param
    }
}


impl soup::DeviceA for Vendor1DeviceA {

    fn op2(&mut self, _: i32) -> i32 {
        self.x += 1;
        self.x
    }
}


impl Vendor1DeviceB {
    fn new(index: i32) -> Vendor1DeviceB {
        Vendor1DeviceB { x: index }
    }
}


impl soup::Device for Vendor1DeviceB {

    fn op1(&mut self, param: i32) {
        self.x *= param
    }
}


impl soup::DeviceB for Vendor1DeviceB {

    fn op2(&mut self, x: i32, y: i32) -> i32 {
        self.x * x + y
    }
}


impl soup::Vendor for Vendor1 {

    fn create_devicea(&self, index: i32) -> Box<soup::DeviceA + 'static> {
        Box::new(Vendor1DeviceA::new(index))
    }

    fn create_deviceb(&self, index: i32) -> Box<soup::DeviceB + 'static> {
        Box::new(Vendor1DeviceB::new(index))
    }
}
